src/Modules/Tags.elm

```diff
@@ -0,0 +1,352 @@
+    }
+
+
+type alias Tag =
+    { name : String
+    , value : String
+    }
+
+
+decoder : Decode.Decoder Tag
+decoder =
+    Decode.succeed Tag
+        |> Decode.required "name" Decode.string
+        |> Decode.optional "value" Decode.string ""
+
+
+initialModel : { tags : Dict String String, isInherited : Bool } -> Model
+initialModel { tags, isInherited } =
+    { tags =
+        tags
+            |> Dict.toList
+            |> List.indexedMap
+                (\index value ->
+                    ( index
+                    , { name = Tuple.first value
+                      , value = Tuple.second value
+                      }
+                    )
+                )
+            |> Dict.fromList
+    , newTagName = ""
+    , newTagValue = ""
+    , isInherited = isInherited
+    , newTagError = Nothing
+    , error = Dict.empty
+    }
+
+
+type Msg
+    = NoOp
+    | NameChanged Int { name : String }
+    | ValueChanged Int { value : String }
+    | Removed Int
+    | TagAdded
+    | NewTagNameChanged String
+    | NewTagValueChanged String
+    | TagValidated Int
+    | Submitted
+    | Closed
+
+
+update : Msg -> Model -> { onSubmit : Model -> msg, onClose : msg, onFocus : msg } -> ( Model, Cmd msg )
+update msg model config =
+    case msg of
+        NoOp ->
+            ( model, Cmd.none )
+
+        NameChanged key { name } ->
+            ( { model
+                | tags =
+                    Dict.update
+                        key
+                        (Maybe.map (\a -> { name = name, value = a.value }))
+                        model.tags
+              }
+            , Cmd.none
+            )
+
+        ValueChanged key { value } ->
+            ( { model
+                | tags =
+                    Dict.update
+                        key
+                        (Maybe.map (\a -> { name = a.name, value = value }))
+                        model.tags
+              }
+            , Cmd.none
+            )
+
+        Removed key ->
+            ( { model
+                | tags = Dict.remove key model.tags
+                , error = Dict.remove key model.error
+              }
+            , Cmd.none
+            )
+
+        TagAdded ->
+            let
+                tagAlreadyExists : Bool
+                tagAlreadyExists =
+                    model.tags
+                        |> Dict.toList
+                        |> List.map (Tuple.second >> .name)
+                        |> List.filter ((==) model.newTagName)
+                        |> List.head
+                        |> Maybe.map (\_ -> True)
+                        |> Maybe.withDefault False
+            in
+            ( if tagAlreadyExists then
+                { model | newTagError = Just "exists" }
+
+              else if String.isEmpty model.newTagName then
+                { model | newTagError = Just "empty" }
+
+              else if String.length model.newTagName > 32 then
+                { model | newTagError = Just "too long (32)" }
```

Comment:
Adding errors to the model, rather than running the validation function with every `view` call, requires us to remember to clear the errors every time.
It's a solution that increases maintenance overhead and exposes us to bugs.

Reply:
I would have something like a "showValidation" flag in the Model, that we set on input blur or (if we really want to be fancy) on a debounced delay after the user has started typing.
Still, good thinking!